File: notifications_backup/src/services/emailService.js
const nodemailer = require('nodemailer');

class EmailService {
  constructor() {
    this.transporter = nodemailer.createTransporter({
      host: process.env.SMTP_HOST || 'smtp.gmail.com',
      port: process.env.SMTP_PORT || 587,
      secure: false, // true for 465, false for other ports
      auth: {
        user: process.env.EMAIL_USER,
        pass: process.env.EMAIL_PASSWORD
      }
    });
  }

  async sendNotificationEmail(notification) {
    const { recipientEmail, title, message, type, priority } = notification;

    const typeIcons = {
      info: '📢',
      warning: '⚠️',
      error: '❌',
      success: '✅'
    };

    const priorityColors = {
      low: '#6B7280',
      medium: '#F59E0B',
      high: '#EF4444'
    };

    const htmlContent = `
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>ProvesiWMS Notification</title>
    </head>
    <body style="margin: 0; padding: 0; font-family: Arial, sans-serif; background-color: #f4f4f4;">
        <table width="100%" cellpadding="0" cellspacing="0" style="background-color: #f4f4f4; padding: 20px;">
            <tr>
                <td align="center">
                    <table width="600" cellpadding="0" cellspacing="0" style="background-color: white; border-radius: 8px; box-shadow: 0 2px 4px rgba(0,0,0,0.1);">
                        <!-- Header -->
                        <tr>
                            <td style="padding: 30px 40px 20px; text-align: center; background-color: #1F2937; border-radius: 8px 8px 0 0;">
                                <h1 style="margin: 0; color: white; font-size: 24px;">ProvesiWMS Notifications</h1>
                            </td>
                        </tr>
                        
                        <!-- Content -->
                        <tr>
                            <td style="padding: 30px 40px;">
                                <div style="border-left: 4px solid ${priorityColors[priority]}; padding-left: 20px; margin-bottom: 20px;">
                                    <h2 style="margin: 0 0 10px 0; color: #1F2937; display: flex; align-items: center;">
                                        <span style="margin-right: 10px; font-size: 24px;">${typeIcons[type]}</span>
                                        ${title}
                                    </h2>
                                    <p style="margin: 0; color: #6B7280; text-transform: uppercase; font-size: 12px; font-weight: bold;">
                                        ${type} • ${priority} priority
                                    </p>
                                </div>
                                
                                <div style="background-color: #F9FAFB; padding: 20px; border-radius: 6px; margin: 20px 0;">
                                    <p style="margin: 0; color: #374151; line-height: 1.6; font-size: 16px;">
                                        ${message}
                                    </p>
                                </div>
                                
                                <div style="margin-top: 30px; padding-top: 20px; border-top: 1px solid #E5E7EB;">
                                    <p style="margin: 0; color: #6B7280; font-size: 14px;">
                                        Sent: ${new Date().toLocaleString()}<br>
                                        Recipient: ${recipientEmail}
                                    </p>
                                </div>
                            </td>
                        </tr>
                        
                        <!-- Footer -->
                        <tr>
                            <td style="padding: 20px 40px; background-color: #F9FAFB; border-radius: 0 0 8px 8px; text-align: center;">
                                <p style="margin: 0; color: #6B7280; font-size: 12px;">
                                    This is an automated notification from ProvesiWMS.<br>
                                    Please do not reply to this email.
                                </p>
                            </td>
                        </tr>
                    </table>
                </td>
            </tr>
        </table>
    </body>
    </html>
    `;

    const mailOptions = {
      from: `"ProvesiWMS Notifications" <${process.env.EMAIL_FROM || process.env.EMAIL_USER}>`,
      to: recipientEmail,
      subject: `${typeIcons[type]} ${title} - ProvesiWMS`,
      html: htmlContent,
      text: `${title}\n\n${message}\n\nType: ${type}\nPriority: ${priority}\nSent: ${new Date().toLocaleString()}`
    };

    try {
      const info = await this.transporter.sendMail(mailOptions);
      console.log(`📧 Email sent to ${recipientEmail}: ${info.messageId}`);
      return {
        success: true,
        messageId: info.messageId,
        sentAt: new Date()
      };
    } catch (error) {
      console.error(`❌ Email failed to ${recipientEmail}:`, error.message);
      return {
        success: false,
        error: error.message,
        sentAt: new Date()
      };
    }
  }

  async testConnection() {
    try {
      await this.transporter.verify();
      console.log('✅ Email service connection verified');
      return true;
    } catch (error) {
      console.error('❌ Email service connection failed:', error.message);
      return false;
    }
  }
}

module.exports = new EmailService();

File: notifications/email_service.py
"""
Email Service for sending notifications via SMTP.

Supports Gmail, Outlook, or any SMTP server.

# TODO: Configurar credenciales SMTP antes de enviar notificaciones
#       Para Gmail:
#       1. Habilitar "Acceso de apps menos seguras" o crear "App Password"
#       2. Ir a https://myaccount.google.com/apppasswords
#       3. Generar una contraseña de aplicación
#       4. Usar esa contraseña en SMTP_PASSWORD
"""
import os
import smtplib
import ssl
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from typing import List

# SMTP Configuration
# Default: Gmail SMTP (can be changed to Outlook, etc.)
SMTP_HOST = os.getenv("SMTP_HOST", "smtp.gmail.com")
SMTP_PORT = int(os.getenv("SMTP_PORT", "587"))
SMTP_USER = os.getenv("SMTP_USER", "")  # Your email address
SMTP_PASSWORD = os.getenv("SMTP_PASSWORD", "")  # App password for Gmail
SENDER_EMAIL = os.getenv("SENDER_EMAIL", "pedropablost@icloud.com")
DEFAULT_RECIPIENT = os.getenv("DEFAULT_RECIPIENT", "p.sanin@uniandes.edu.co")

# Email sending enabled flag (disable if no SMTP configured)
EMAIL_ENABLED = os.getenv("EMAIL_ENABLED", "false").lower() == "true"


async def send_notification_email(
    recipient_email: str,
    subject: str,
    message: str,
    sender_username: str
) -> bool:
    """
    Send a notification email via SMTP.
    
    Args:
        recipient_email: Email address to send to
        subject: Email subject
        message: Notification message
        sender_username: Username of who sent the notification
    
    Returns:
        True if email sent successfully, False otherwise
    """
    # Check if email is enabled
    if not EMAIL_ENABLED:
        print(f"📧 Email disabled. Would send to: {recipient_email}")
        print(f"   Subject: {subject}")
        print(f"   Message: {message}")
        return True  # Return True to not block the notification flow
    
    # Check SMTP credentials
    if not SMTP_USER or not SMTP_PASSWORD:
        print("⚠️  SMTP credentials not configured. Email not sent.")
        return False
    
    # Use default recipient for testing
    actual_recipient = recipient_email or DEFAULT_RECIPIENT
    
    try:
        # Create message
        msg = MIMEMultipart("alternative")
        msg["Subject"] = f"[ProvesiWMS] {subject}"
        msg["From"] = SENDER_EMAIL
        msg["To"] = actual_recipient
        
        # Plain text version
        text_content = f"""
Nueva notificación de ProvesiWMS
================================

De: {sender_username}
Mensaje: {message}

---
Este es un mensaje automático del sistema de notificaciones de ProvesiWMS.
        """
        
        # HTML version
        html_content = f"""
<!DOCTYPE html>
<html>
<head>
    <style>
        body {{ font-family: Arial, sans-serif; margin: 0; padding: 20px; background-color: #f4f4f4; }}
        .container {{ max-width: 600px; margin: 0 auto; background: white; padding: 20px; border-radius: 8px; box-shadow: 0 2px 4px rgba(0,0,0,0.1); }}
        .header {{ background: #2196F3; color: white; padding: 15px; border-radius: 8px 8px 0 0; margin: -20px -20px 20px -20px; }}
        .message {{ background: #f9f9f9; padding: 15px; border-left: 4px solid #2196F3; margin: 15px 0; }}
        .footer {{ color: #666; font-size: 12px; margin-top: 20px; padding-top: 15px; border-top: 1px solid #eee; }}
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h2 style="margin: 0;">📬 Nueva Notificación</h2>
        </div>
        <p><strong>De:</strong> {sender_username}</p>
        <div class="message">
            <p>{message}</p>
        </div>
        <div class="footer">
            <p>Este es un mensaje automático del sistema de notificaciones de ProvesiWMS.</p>
        </div>
    </div>
</body>
</html>
        """
        
        # Attach both versions
        msg.attach(MIMEText(text_content, "plain"))
        msg.attach(MIMEText(html_content, "html"))
        
        # Create secure SSL context
        context = ssl.create_default_context()
        
        # Connect and send
        with smtplib.SMTP(SMTP_HOST, SMTP_PORT) as server:
            server.starttls(context=context)
            server.login(SMTP_USER, SMTP_PASSWORD)
            server.sendmail(SENDER_EMAIL, actual_recipient, msg.as_string())
        
        print(f"✅ Email sent to {actual_recipient}!")
        return True
        
    except smtplib.SMTPAuthenticationError as e:
        print(f"❌ SMTP Authentication failed: {str(e)}")
        print("   Tip: For Gmail, use an App Password instead of your regular password")
        return False
    except smtplib.SMTPException as e:
        print(f"❌ SMTP error: {str(e)}")
        return False
    except Exception as e:
        print(f"❌ Unexpected error sending email: {str(e)}")
        return False


async def send_bulk_notification_emails(
    recipients: List[str],
    subject: str,
    message: str,
    sender_username: str
) -> dict:
    """
    Send notification to multiple recipients.
    
    Returns:
        Dict with 'success' and 'failed' counts
    """
    results = {"success": 0, "failed": 0}
    
    for recipient in recipients:
        if await send_notification_email(recipient, subject, message, sender_username):
            results["success"] += 1
        else:
            results["failed"] += 1
    
    return results
